py: Add ustruct.pack_into and unpack_from

dhylands opened the pull request:
Having pack_into and unpack_from allows allocating fewer buffers.

dhylands committed:
py: Add ustruct.pack_into and unpack_from

pfalcon commented:
Fewer, but not exactly zero. And this same trait (fewer buffers) is achievable by uctypes. So, what improvement this patch gives? How many bytes does it cost?

dhylands commented:
> Fewer, but not exactly zero. And this same trait (fewer buffers) is achievable by uctypes. So, what improvement this patch gives? How many bytes does it cost?

Well you need the underlying buffer object, but that's true for uctypes as well. I might be able to use uctypes, but it would a bunch more code required, since I'm using this to unpack packets over the uart, and the packet layout varies from packet to packet (and no the packet format is not controllable by me - I'm talking to Bioloid AX-12 servos, and I can't change the firmware in the servos.

I might be able to use uctypes - not sure. Using uctypes will cost me RAM to store the structures. Using pack_into and unpack_from costs me flash (and a bit of RAM for the strings which describe the packing format).

This adds about 400 bytes of flash to implement both pack_into and unpack_from (a big chunk of that is for the parameter checking and the exceptions).

I'm also happy to make these configable. I saw that there was a config option for struct as a whole. I'd be happy to make them individually configable, or as a pair (and perhaps even make pack and unpack configable.

dpgeorge commented:
To reduce size you can make unpack and unpack_from the same function.  They have effectively equivalent type signatures.

Also, make the exception messages shorter and the same for both functions would help, something like "buffer too short" would be enough.

dhylands commented:
Sounds reasonable. unpack says that the buffer has to be exactly the right size, and unpack_from says that it just has to be "big enough". So as long as we're willing to relax the "exactly" portion of the specification, then we could make them be the same function.

Or were you thinking of something slightly different?

pfalcon commented:
> So as long as we're willing to relax the "exactly" portion of the specification, then we could make them be the same function.

Yes, that's ok, we have relaxed rules in many places in uPy.

pfalcon commented:
> This adds about 400 bytes of flash to implement both pack_into and unpack_from (a big chunk of that is for the parameter checking and the exceptions).

> I'm also happy to make these configable.

So, 400 bytes of Thumb2 code is quite large, and thinking how to reduce that, along @dpgeorge's lines, makes sense. It may also make sense to still make additions configurable, but let's leave that for next iteration of this patch.

dpgeorge commented:
> And this same trait (fewer buffers) is achievable by uctypes. So, what improvement this patch gives?

Just to give my opinion on this: while I agree it's not great to duplicate functionality (although Python does like to do that, we can all give countless examples) modstruct is in the core but moductypes is in extmod, so it's highly reasonable that a given port with not much flash space would disable moductypes but keep modstruct so that it has at least some binary functionality.  Thus it's a good idea to support these RAM saving methods in modstruct.

dhylands commented:
Reworked, as per suggestions. Size increase on stmhal is now only 260 bytes.
struct.unpack now raises an exception if the provided buffer is too small.

dpgeorge commented on py/modstruct.c:
```diff
@@ -132,17 +132,34 @@ STATIC mp_obj_t struct_calcsize(mp_obj_t fmt_in) {
+    // unpack_from requires that the buffer be "big enough".
+    // Since we implement unpack and unpack_from using the same function
+    // we relax the "exact" requirement, and only implement "big enough".
+    uint size = MP_OBJ_SMALL_INT_VALUE(struct_calcsize(args[0]));
```
I can't help but think that it's inefficient to call struct_calcsize, and then calcsize_items below, and then parse the fmt a third time... maybe the check for "buffer too small" can be done in the final loop that parses fmt?

dhylands replied:
Perhaps it would make sense to have a function that calculates both the number of items and the number of bytes? It isn't obvious from the names, but struct_calcsize calculates the number of bytes, and calcsize_items calculate the number of items.

It should be possible to do the check in the loop as well. I'll see how that affects things (and it should be possible to do this for both pack and unpack).

dpgeorge replied:
@dhylands were you going to look further into improving this?

dhylands replied:
> @dhylands were you going to look further into improving this?
@dpgeorge I did, and forgot to push the changes - Doh

Now pushed.

dpgeorge commented on py/modstruct.c:
```diff
@@ -163,23 +183,20 @@ STATIC mp_obj_t struct_unpack(mp_obj_t fmt_in, mp_obj_t data_in) {
-    vstr_t vstr;
-    vstr_init_len(&vstr, size);
-    byte *p = (byte*)vstr.buf;
-    memset(p, 0, size);
```
This memset seems to have disappeared... I guess it's original reason for being was to make sure that if you didn't pass enough args for the pack spec then they were at least zero in the returned bytes object.

Do we still need/want it?  For struct.pack at least?

dhylands commented:
I added the code to do the checking to verify that the number of arguments matches the format.

I added back the memset, since even if we detect a mismatch we still probably want the bytes zero'd to prevent some weird GC issue.

I added tests to check the too many/too few arguments.

The patch now increases stmhal by 332 bytes.

dpgeorge commented:
> I added the code to do the checking to verify that the number of arguments matches the format.

I wasn't asking for this, and I don't think we need it.  If simpler (and smaller) code just so happens to be a superset of CPython's behaviour then that's fine for us.

dpgeorge commented:
I removed the check for too many / too few args and merged in a17755ee8ba2982505d349e0f06acdabe943d33e.

Stmhal increase is 284 bytes.
